### PR TITLE
Fixes for encoder abstraction.

### LINKED ISSUE #23191: [Bug] Encoder on slave continues to trigger forever once activated
### Describe the Bug

After updating from an older version to latest git and flashing both halves of my keyboard, the non-primary side encoder, once rotated continues to trigger forever at a rate that causes the keyboard to become unresponsive.

The primary side operates as you might expect. 

I've narrowed it down to commit 9d9cdaaa2d035787b0b50c26f2975695fdbc16f4.
The commit before, 2eb9ff8efd1df2c98724481c71c8ab8a5b62e31e functions OK.

### Keyboard Used

sofle_choc

### Link to product page (if applicable)

Linux

### Operating System

_No response_

### qmk doctor Output

```text
QMK Doctor is checking your environment.
Ψ CLI version: 1.1.2
Ψ QMK home: /home/jason/keyboards/qmk/master
Ψ Detected Linux (Fedora Linux 39 (Workstation Edition)).
Ψ Userspace enabled: False
Ψ Git branch: master
Ψ Repo version: 0.24.0
Ψ - Latest master: 2024-02-28 21:47:37 +1100 (4e369d405af) -- Merge branch 'develop'
Ψ - Latest upstream/master: 2024-02-28 21:47:37 +1100 (4e369d405af) -- Merge branch 'develop'
Ψ - Latest upstream/develop: 2024-02-28 19:13:04 -0700 (4e953f11696) -- Fixup mechlovin/octagon (#23179)
Ψ - Common ancestor with upstream/master: 2024-02-28 21:47:37 +1100 (4e369d405af) -- Merge branch 'develop'
Ψ - Common ancestor with upstream/develop: 2024-02-28 21:47:37 +1100 (4e369d405af) -- Merge branch 'develop'
Ψ All dependencies are installed.
Ψ Found arm-none-eabi-gcc version 13.1.0
Ψ Found avr-gcc version 13.1.0
⚠ We do not recommend avr-gcc newer than 8. Downgrading to 8.x is recommended.
Ψ Found avrdude version 7.3
Ψ Found dfu-programmer version 0.7.2
Ψ Found dfu-util version 0.11
Ψ Submodules are up to date.
Ψ Submodule status:
Ψ - lib/chibios: 2023-04-15 13:48:04 +0000 --  (11edb1610)
Ψ - lib/chibios-contrib: 2023-11-27 18:15:44 +0100 --  (9d7a7f90)
Ψ - lib/googletest: 2021-06-11 06:37:43 -0700 --  (e2239ee6)
Ψ - lib/lufa: 2022-08-26 12:09:55 +1000 --  (549b97320)
Ψ - lib/vusb: 2022-06-13 09:18:17 +1000 --  (819dbc1)
Ψ - lib/printf: 2022-06-29 23:59:58 +0300 --  (c2e3b4e)
Ψ - lib/pico-sdk: 2023-02-12 20:19:37 +0100 --  (a3398d8)
Ψ - lib/lvgl: 2022-04-11 04:44:53 -0600 --  (e19410f)
Ψ QMK is ready to go, but minor problems were found
```

### Is AutoHotKey / Karabiner installed

- [ ] AutoHotKey (Windows)
- [ ] Karabiner (macOS)

### Other keyboard-related software installed

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by tzarc**
## Description

As reported on Discord ([here onwards](https://discord.com/channels/440868230475677696/440868230475677698/1212999470619566100)), and subsequently #23191.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* (Hopefully) Fixes #23191.

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by jasonish** [comment]
As mentioned in Discord, this didn't help. I might poke around  bit myself, but not sure if I'll get anywhere. But happy test anything.  Just @ me or I'll keep an eye on this pull request.

**Comment by jasonish** [comment]
This is working for me so far, the second encoder is now usable.

**Comment by ireun** [comment]
I can also confirm it's working - Sofle RGB here

**Comment by Coko7** [comment]
As mentioned in my [comment](https://github.com/qmk/qmk_firmware/issues/23191#issuecomment-1982322838) on issue #23191, this fix solves the problem for me, and I have yet to face the initial problem again.
I am using a `sofle/rev1` with RGB and I confirm that it is working.

**Comment by tzarc** [comment]
@finrod09 @jasonish @ireun @Coko7 once more a set of changes, would be great to confirm.

Marking it as ready for review in the meantime, hopefully if it all goes well on your end we can sort out a merge on ours.

File: quantum/encoder.c [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include "action.h"
 #include "encoder.h"
 #include "wait.h"
+#include "debug.h"
```
Comment (tzarc) [inline quantum/encoder.c:8]: Remove debugging

File: quantum/encoder.c [hunk 2]
```diff
@@ -15,34 +16,40 @@ __attribute__((weak)) bool should_process_encoder(void) {
...
     encoder_driver_init();
 }
 
-static bool encoder_handle_queue(void) {
-    bool changed = false;
-    while (encoder_events.tail != encoder_events.head) {
-        encoder_event_t event = encoder_events.queue[encoder_events.tail];
-        encoder_events.tail   = (encoder_events.tail + 1) % MAX_QUEUED_ENCODER_EVENTS;
+static void encoder_queue_drain(void) {
+    // dprintf("Draining encoder queue\n");
```
Comment (tzarc) [inline quantum/encoder.c:27]: Remove debugging

File: quantum/encoder.c [hunk 3]
```diff
@@ -72,39 +84,75 @@ bool encoder_task(void) {
...
-    encoder_events.queue[encoder_events.head] = new_event;
+    encoder_event_t new_event   = {.index = index, .clockwise = clockwise ? 1 : 0};
+    events->queue[events->head] = new_event;
 
     // Increment the head index
-    encoder_events.head = (encoder_events.head + 1) % MAX_QUEUED_ENCODER_EVENTS;
+    events->head = (events->head + 1) % MAX_QUEUED_ENCODER_EVENTS;
+    events->enqueued++;
+
+    // dprintf("Enqueued event: %d %s [h=%d,t=%d,e=%d,d=%d]\n", (int)index, clockwise ? "CW" : "CCW", (int)events->head, (int)events->tail, (int)events->enqueued, (int)events->dequeued);
```
Comment (tzarc) [inline quantum/encoder.c:117]: Remove debugging

File: quantum/encoder.c [hunk 4]
```diff
@@ -72,39 +84,75 @@ bool encoder_task(void) {
...
+    // Retrieve the event
+    encoder_event_t event = events->queue[events->tail];
+    *index                = event.index;
+    *clockwise            = event.clockwise;
+
+    // Increment the tail index
+    events->tail = (events->tail + 1) % MAX_QUEUED_ENCODER_EVENTS;
+    events->dequeued++;
+
+    // dprintf("Dequeued event: %d %s [h=%d,t=%d,e=%d,d=%d]\n", (int)*index, *clockwise ? "CW" : "CCW", (int)events->head, (int)events->tail, (int)events->enqueued, (int)events->dequeued);
```
Comment (tzarc) [inline quantum/encoder.c:136]: Remove debugging

**Review by tzarc — COMMENTED** [review]
Notes for self

**Comment by ireun** [comment]
Still fine on my end, encoders on both sides of Sofle RGB working correctly. Thank You @tzarc. <3
```
Ψ Git branch: bugfix/encoder-abstraction
Ψ Repo version: 0.24.2
Ψ - Latest bugfix/encoder-abstraction: 2024-03-08 23:00:42 +1100 (d4eda060e3) -- Move queue drain to main thread.
```

**Review by drashna — APPROVED** [review]

**Pull request merged**